Add trajectory transformer

CarlCochet opened the pull request:
# What does this PR do?
<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

Add Trajectory Transformer to transformers.

This is a transformer model used for offline deep reinforcement learning, taking observations, actions and rewards as one big sequence. 
The model is adapted from the code [here](https://github.com/jannerm/trajectory-transformer). 
The caching mechanism as well as attentions outputs and hidden states outputs are additions not present in the original code. 

The adapted forward pass produces the same output tensors as the forward pass in the original model.

Example usage for random environment data:

```python
from transformers import TrajectoryTransformerModel
import torch
import numpy as np

config = TrajectoryTransformerConfig.from_pretrained("CarlCochet/trajectory-transformer-halfcheetah-medium-v2")
model = TrajectoryTransformerModel.from_pretrained("CarlCochet/trajectory-transformer-halfcheetah-medium-v2", config=config)

batch_size = 1
# sequence made of [ actions, observations, reward ]
seq_length = config.action_dim + config.observation_dim + 1 

trajectories = torch.LongTensor([np.random.permutation(seq_length) for _ in range(batch_size)])
outputs = model(trajectories)

prediction_logits = outputs.logits
```

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## TODO

- Pass all relevant tests ; need to disable some irrelevant tests as well
- Make a demo notebook

## Who can review?

Anyone in the community is free to review the PR once the tests have passed.
@LysandreJik @thomwolf @edbeeching 

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- albert, bert, xlm: @LysandreJik
- blenderbot, bart, marian, pegasus, encoderdecoder,  t5: @patrickvonplaten, @patil-suraj
- longformer, reformer, transfoxl, xlnet: @patrickvonplaten
- fsmt: @stas00
- funnel: @sgugger
- gpt2: @patrickvonplaten, @LysandreJik
- rag: @patrickvonplaten, @lhoestq
- tensorflow: @LysandreJik

Library:

- benchmarks: @patrickvonplaten
- deepspeed: @stas00
- ray/raytune: @richardliaw, @amogkam
- text generation: @patrickvonplaten
- tokenizers: @n1t0, @LysandreJik
- trainer: @sgugger
- pipelines: @LysandreJik

Documentation: @sgugger

HF projects:

- datasets: [different repo](https://github.com/huggingface/datasets)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Examples:

- maintained examples (not research project or legacy): @sgugger, @patil-suraj
- research_projects/bert-loses-patience: @JetRunner
- research_projects/distillation: @VictorSanh

 -->

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_17141). All of your documentation changes will be reflected on that endpoint.

LysandreJik commented on src/transformers/models/trajectory_transformer/__init__.py:
```diff
@@ -0,0 +1,68 @@
+# There's no way to ignore "F401 '...' imported but unused" warnings in this
+# module, but to preserve other warnings. So, don't check this module at all.
+
+# Copyright 2020 The HuggingFace Team. All rights reserved.
```
We're in 2022 :)

CarlCochet commented on .gitattributes:
```diff
@@ -1,3 +1,4 @@
-*.md	eol=lf
\ No newline at end of file
+*.md	eol=lf
+*.mdx   eol=lf
```
Needed to add this for Windows because fix-copies and fixup were incorrectly modifying index.mdx and serialization.mdx (in docs/source/en)

LysandreJik commented on src/transformers/models/trajectory_transformer/configuration_trajectory_transformer.py:
```diff
@@ -0,0 +1,167 @@
+logger = logging.get_logger(__name__)
+
+TRAJECTORY_TRANSFORMER_PRETRAINED_CONFIG_ARCHIVE_MAP = {
+    "CarlCochet/trajectory-transformer-halfcheetah-medium-v2": "https://huggingface.co/CarlCochet/trajectory-transformer-halfcheetah-medium-v2/resolve/main/config.json",
```
Before merging, it would be nice to upload these to the author's organization. Have you been in contact with the authors? It would be nice to ask them if they want to have an organization on the hub and to host these weights there.

Let me know if you'd like some pointers.

CarlCochet replied:
I have indeed been in contact with the authors and could probably ask if they want to do that, however their pretrained checkpoints are not compatible with this model. I had to modify them on my end to make them compatible with transformers, so I'm not sure how that would work.

LysandreJik commented on src/transformers/models/trajectory_transformer/configuration_trajectory_transformer.py:
```diff
@@ -0,0 +1,167 @@
+        vocab_size (`int`, *optional*, defaults to 100):
+            Vocabulary size of the TrajectoryTransformer model. Defines the number of different tokens that can be
+            represented by the `trajectories` passed when calling [`TrajectoryTransformerModel`] or
+            [`~TFTrajectoryTransformerModel`].
```
This model doesn't exist

LysandreJik commented on src/transformers/models/trajectory_transformer/configuration_trajectory_transformer.py:
```diff
@@ -0,0 +1,167 @@
+            just in case (e.g., 512 or 1024 or 2048).
+        type_vocab_size (`int`, *optional*, defaults to 2):
+            The vocabulary size of the `token_type_ids` passed when calling [`TrajectoryTransformerModel`] or
+            [`~TFTrajectoryTransformerModel`].
```
Same here, TF version doesn't exist

LysandreJik commented on src/transformers/models/trajectory_transformer/modeling_trajectory_transformer.py:
```diff
@@ -0,0 +1,604 @@
+    base_model_prefix = "trajectory_transformer"
+    main_input_name = "trajectories"
+    supports_gradient_checkpointing = True
+    _keys_to_ignore_on_load_missing = [r"position_ids"]
```
I don't think this line is necessary

CarlCochet replied:
You mean the `_keys_to_ignore_on_load_missing = [r"position_ids"]` line?

sgugger replied:
Yes, it doesn't look like the model is defining this attribute.

CarlCochet replied:
Indeed, I'll remove it.

LysandreJik commented on src/transformers/models/trajectory_transformer/modeling_trajectory_transformer.py:
```diff
@@ -0,0 +1,604 @@
+    behavior.
+
+    Parameters:
+        config ([`~TrajectoryTransformerConfig`]): Model configuration class with all the parameters of the model.
```
```suggestion
        config ([`TrajectoryTransformerConfig`]): Model configuration class with all the parameters of the model.
```

LysandreJik commented on src/transformers/models/trajectory_transformer/modeling_trajectory_transformer.py:
````diff
@@ -0,0 +1,604 @@
+
+        Examples:
+
+        ```python
+        >>> from transformers import TrajectoryTransformerModel
+        >>> import torch
+
+        >>> model = TrajectoryTransformerModel.from_pretrained(
+        ...     "CarlCochet/trajectory-transformer-halfcheetah-medium-v2"
+        ... )
+        >>> # evaluation
+        >>> model = model.to(device)
+        >>> model.eval()
+
+        >>> logits, loss = model(trajectories, **kwargs)
+        ```
````
Reading this, I'm wondering what are the `kwargs` and `trajectories`. I'd favor a more complete code example here; but if `trajectories` are hard to obtain, then providing a randomly initialized `trajectories` tensor that would show the expected shapes, alongside a link to a colab going into detail regarding how to obtain those trajectories would help.

CarlCochet replied:
So, I updated this doc to be a more complete example but it's using a randomly generated tensor for trajectories. Where should I put a link to a collab to show a more complete example? Here or in the .mdx docs?

LysandreJik commented on tests/models/trajectory_transformer/test_modeling_trajectory_transformer.py:
```diff
@@ -0,0 +1,279 @@
+    # # Input is 'trajectories' not 'input_ids'
+    def test_model_main_input_name(self):
+        model_signature = inspect.signature(getattr(TrajectoryTransformerModel, "forward"))
+        print(model_signature)
```
No need for the print statement

LysandreJik commented on tests/models/trajectory_transformer/test_modeling_trajectory_transformer.py:
```diff
@@ -0,0 +1,279 @@
+        print(model_signature)
+        # The main input is the name of the argument after `self`
+        observed_main_input_name = list(model_signature.parameters.keys())[1]
+        print(observed_main_input_name, TrajectoryTransformerModel.main_input_name)
```
same here

LysandreJik commented on tests/models/trajectory_transformer/test_modeling_trajectory_transformer.py:
```diff
@@ -0,0 +1,279 @@
+class TrajectoryTransformerModelIntegrationTest(unittest.TestCase):
+    @slow
+    def test_prediction(self):
+        np.random.seed(42)
+        torch.manual_seed(42)
```
Have you found this necessary in your tests?

LysandreJik commented on src/transformers/models/trajectory_transformer/modeling_trajectory_transformer.py:
```diff
@@ -0,0 +1,621 @@
+    Args:
+        trajectories (`torch.LongTensor` of shape `(batch_size, sequence_length)`):
+            Batch of trajectories, where a trajectory is a sequence of states, actions and rewards.
+        past_key_values (`Tuple[Tuple[torch.Tensor]]` of length `config.n_layers`):
+            Contains precomputed hidden-states (key and values in the attention blocks) as computed by the model (see
+            `past_key_values` output below). Can be used to speed up sequential decoding. The `input_ids` which have
+            their past given to this model should not be passed as `input_ids` as they have already been computed.
+        targets (`torch.LongTensor` of shape `(batch_size, sequence_length)`):
```
I think these should be marked as `*optional*` as well.

LysandreJik commented on docs/source/en/model_doc/trajectory_transformer.mdx:
```diff
@@ -0,0 +1,49 @@
+specific language governing permissions and limitations under the License.
+-->
+
+# TrajectoryTransformer
```
Given that it's a brand new model with an API that is very different from the original implementation, I think it would significantly help usage if we could add some in-depth examples here, either in the markdown or as a link to a colab notebook.

LysandreJik reviewed (commented):
This looks quite good already! I have only left nits. I think an in-depth example would really help comprehension.

sgugger commented on src/transformers/models/trajectory_transformer/modeling_trajectory_transformer.py:
```diff
@@ -0,0 +1,621 @@
+        else:
+            present = None
+
+        # causal self-attention; Self-attend: (B, nh, T, hs) x (B, nh, hs, T) -> (B, nh, T, T)
+        # [ B x n_heads x T x T ]
+        attn_weights = (torch.matmul(query, key.transpose(-2, -1))) * (1.0 / math.sqrt(key.size(-1)))
+        attn_weights = attn_weights.masked_fill(
+            self.mask[:, :, :sequence_length, :sequence_length] == 0, float("-inf")
+        )
+        attn_weights = F.softmax(attn_weights, dim=-1)
+        self._attn_map = attn_weights.clone()
+        attn_weights = self.attn_drop(attn_weights)
+
+        # [ B x n_heads x T x head_size ]
+        output = torch.matmul(attn_weights, value)  # (B, nh, T, T) x (B, nh, T, hs) -> (B, nh, T, hs)
+        # [ B x T x embedding_dim ]
```
Same thing here, expand one-letter names

sgugger commented on src/transformers/models/trajectory_transformer/modeling_trajectory_transformer.py:
```diff
@@ -0,0 +1,621 @@
+        if output_attentions:
+            outputs += (attn_weights,)
+
+        return outputs  # attn_output, present, (attn_weights)
```
```suggestion
        return outputs
```
To clean up?

sgugger commented on src/transformers/models/trajectory_transformer/modeling_trajectory_transformer.py:
```diff
@@ -0,0 +1,621 @@
+        self.blocks = nn.ModuleList([Block(config) for _ in range(config.n_layer)])
+        # decoder head
+        self.ln_f = nn.LayerNorm(config.n_embd)
+        # self.head = nn.Linear(config.n_embd, config.vocab_size, bias=False)
```
```suggestion
```

sgugger commented on src/transformers/models/trajectory_transformer/modeling_trajectory_transformer.py:
```diff
@@ -0,0 +1,621 @@
+            raise ValueError("Cannot forward, model block size is exhausted.")
+
+        offset_trajectories = self.offset_tokens(trajectories)
+        # [ B x T x embedding_dim ]
+        # forward the GPT model
+        token_embeddings = self.tok_emb(offset_trajectories)  # each index maps to a (learnable) vector
+        # [ 1 x T x embedding_dim ]
+        position_embeddings = self.pos_emb[:, :sequence_length, :]  # each position maps to a (learnable) vector
+        # [ B x T x embedding_dim ]
```
Same thing here.

sgugger commented on src/transformers/models/trajectory_transformer/modeling_trajectory_transformer.py:
```diff
@@ -0,0 +1,621 @@
+                    output_attentions,
+                )
+            else:
+                outputs = block(
+                    hidden_states,
+                    layer_past,
+                    use_cache,
+                    output_attentions,
+                )
```
```suggestion
                outputs = block(hidden_states, layer_past, use_cache, output_attentions)
```
Fits in one line.

sgugger reviewed (approved):
Thanks a lot for adding this new model!

CarlCochet committed:
Add trajectory transformer


Fix model init


Fix end of lines for .mdx files

Add trajectory transformer model to toctree

Add forward input docs

Fix docs, remove prints, simplify prediction test

Apply suggestions from code review

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>
Apply suggestions from code review

Co-authored-by: Lysandre Debut <lysandre@huggingface.co>
Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>
Update docs, more descriptive comments

Apply suggestions from code review

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>
Update readme

Small comment update and add conversion script

Rebase and reformat

Fix copies

Fix rebase, remove duplicates

Fix rebase, remove duplicates

CarlCochet committed:
Remove tapex

CarlCochet committed:
Remove tapex

CarlCochet committed:
Remove tapex

LysandreJik reviewed (approved):
LGTM, thanks @CarlCochet!
